tests/testdiscovery.nim

```diff
@@ -32,32 +32,119 @@ proc createSwitch(rdv: RendezVous = RendezVous.new()): Switch =
   .build()
 
 suite "Discovery":
-  teardown:
-    checkTrackers()
-  asyncTest "RendezVous test":
-    let
-      rdvA = RendezVous.new()
-      rdvB = RendezVous.new()
-      clientA = createSwitch(rdvA)
-      clientB = createSwitch(rdvB)
-      remoteNode = createSwitch()
-      dmA = DiscoveryManager()
-      dmB = DiscoveryManager()
+  const
+    namespace = "foo"
+    rdvNamespace = RdvNamespace(namespace)
+
+  var
+    rdvA {.threadvar.}: RendezVous
+    rdvB {.threadvar.}: RendezVous
+    rdvRemote {.threadvar.}: RendezVous
+    clientA {.threadvar.}: Switch
+    clientB {.threadvar.}: Switch
+    remoteNode {.threadvar.}: Switch
+    dmA {.threadvar.}: DiscoveryManager
+    dmB {.threadvar.}: DiscoveryManager
+
+  asyncSetup:
+    rdvA = RendezVous.new()
+    rdvB = RendezVous.new()
+    rdvRemote = RendezVous.new()
+    clientA = createSwitch(rdvA)
+    clientB = createSwitch(rdvB)
+    remoteNode = createSwitch(rdvRemote)
+    dmA = DiscoveryManager()
+    dmB = DiscoveryManager()
+
     dmA.add(RendezVousInterface.new(rdvA, ttr = 500.milliseconds))
     dmB.add(RendezVousInterface.new(rdvB))
     await allFutures(clientA.start(), clientB.start(), remoteNode.start())
 
     await clientB.connect(remoteNode.peerInfo.peerId, remoteNode.peerInfo.addrs)
     await clientA.connect(remoteNode.peerInfo.peerId, remoteNode.peerInfo.addrs)
 
-    dmB.advertise(RdvNamespace("foo"))
+  asyncTeardown:
+    await allFutures(clientA.stop(), clientB.stop(), remoteNode.stop())
+    checkTrackers()
+
+  asyncTest "RendezVous test":
+    dmB.advertise(rdvNamespace)
 
     let
-      query = dmA.request(RdvNamespace("foo"))
+      query = dmA.request(rdvNamespace)
       res = await query.getPeer()
     check:
       res{PeerId}.get() == clientB.peerInfo.peerId
       res[PeerId] == clientB.peerInfo.peerId
       res.getAll(PeerId) == @[clientB.peerInfo.peerId]
       toHashSet(res.getAll(MultiAddress)) == toHashSet(clientB.peerInfo.addrs)
-    await allFutures(clientA.stop(), clientB.stop(), remoteNode.stop())
+
+  asyncTest "Subscribe and unsubscribe":
+    dmB.advertise(rdvNamespace)
+    let
+      query1 = dmA.request(rdvNamespace)
+      res1 = await query1.getPeer().waitForResult(1.seconds)
+
+    res1.assertIsOk()
+    check res1.value{PeerId}.get() == clientB.peerInfo.peerId
+
+    await rdvB.unsubscribe(namespace)
+    var
+      query2 = dmA.request(rdvNamespace)
+      res2 = await query2.getPeer().waitForResult(1.seconds)
+
+    res2.assertIsErr()
+
+  asyncTest "Frequent sub/unsub":
+    for i in 0 ..< 10:
+      dmB.advertise(rdvNamespace)
+      let
+        query1 = dmA.request(rdvNamespace)
+        res1 = await query1.getPeer().waitForResult(1.seconds)
```

Comment:
A `query` is a tool to re-use. Recreating one every loop is not the way to do.
It should work just by writing
```nim
asyncTest "Frequent sub/unsub":
  let
    query1 = dmA.request(rdvNamespace)
    query2 = dmA.request(rdvNamespace)
  for i in 0 ..< 10:
    (...)
```

Reply:
Just did that and it doesn't work, it breaks at the 3rd iteration. I don't know what might the correlation be.

Reply:
Then there's a problem in rendezvous or the discovery manager :confused:

Reply:
What would you rather I do with this? Leave it as is and open an issue or would you like to investigate it before merging?

Reply:
I would like to investigate, it's a real problem. I'll do this next week.